sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTrackerTest.java

```diff
@@ -0,0 +1,196 @@
+    tracker.tryClaim(ByteKey.of(0x70));
+  }
+
+  @Test
+  public void testClaimBeforeStartOfRange() throws Exception {
+    expected.expectMessage(
+        "Trying to claim key [05] before start of the range "
+            + "ByteKeyRange{startKey=[10], endKey=[c0]}");
+    ByteKeyRangeTracker tracker =
+        ByteKeyRangeTracker.of(ByteKeyRange.of(ByteKey.of(0x10), ByteKey.of(0xc0)));
+    tracker.tryClaim(ByteKey.of(0x05));
+  }
+
+  @Test
+  public void testCheckDoneAfterTryClaimPastEndOfRange() {
+    ByteKeyRangeTracker tracker =
+        ByteKeyRangeTracker.of(ByteKeyRange.of(ByteKey.of(0x10), ByteKey.of(0xc0)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0x50)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0x90)));
+    assertFalse(tracker.tryClaim(ByteKey.of(0xd0)));
+    tracker.checkDone();
+  }
+
+  @Test
+  public void testCheckDoneAfterTryClaimAtEndOfRange() {
+    ByteKeyRangeTracker tracker =
+        ByteKeyRangeTracker.of(ByteKeyRange.of(ByteKey.of(0x10), ByteKey.of(0xc0)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0x50)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0x90)));
+    assertFalse(tracker.tryClaim(ByteKey.of(0xc0)));
+    tracker.checkDone();
+  }
+
+  @Test
+  public void testCheckDoneAfterTryClaimRightBeforeEndOfRange() {
+    ByteKeyRangeTracker tracker =
+        ByteKeyRangeTracker.of(ByteKeyRange.of(ByteKey.of(0x10), ByteKey.of(0xc0)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0x50)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0x90)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0xbf)));
+    tracker.checkDone();
+  }
+
+  @Test
+  public void testCheckDoneWhenNotDone() {
+    ByteKeyRangeTracker tracker =
+        ByteKeyRangeTracker.of(ByteKeyRange.of(ByteKey.of(0x10), ByteKey.of(0xc0)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0x50)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0x90)));
+    expected.expectMessage(
+        "Last attempted key was [90] in range ByteKeyRange{startKey=[10], endKey=[c0]}, "
+            + "claiming work in [[91], [c0]) was not attempted");
+    tracker.checkDone();
+  }
+
+  @Test
+  public void testCheckDoneWhenExplicitlyMarkedDone() {
+    ByteKeyRangeTracker tracker =
+        ByteKeyRangeTracker.of(ByteKeyRange.of(ByteKey.of(0x10), ByteKey.of(0xc0)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0x50)));
+    assertTrue(tracker.tryClaim(ByteKey.of(0x90)));
+    tracker.markDone();
+    tracker.checkDone();
+  }
+
+  @Test
+  public void testNextByteKey() {
+    assertEquals(next(ByteKey.EMPTY), ByteKey.of(0x00));
+    assertEquals(next(ByteKey.of(0x00)), ByteKey.of(0x01));
+    assertEquals(next(ByteKey.of(0x9f)), ByteKey.of(0xa0));
+    assertEquals(next(ByteKey.of(0xff)), ByteKey.of(0x01, 0x00));
+    assertEquals(next(ByteKey.of(0x10, 0x10)), ByteKey.of(0x10, 0x11));
+    assertEquals(next(ByteKey.of(0x00, 0xff)), ByteKey.of(0x01, 0x00));
+    assertEquals(next(ByteKey.of(0xff, 0xff)), ByteKey.of(0x01, 0x00, 0x00));
+  }
```

Comment:
Are there tests for corner cases like "checkpointing produces an empty residual range" (but doesn't fail), "tryClaim in an empty range doesn't throw" (but returns false)?

Reply:
Yes `testCheckpointClaimedLast` produces an empty range and does not fail. And `testTryClaim` covers the second case.